x-pack/plugin/profiling/src/main/java/org/elasticsearch/xpack/profiling/InstanceType.java

```diff
@@ -45,16 +45,47 @@ public static InstanceType fromCostSource(Map<String, Object> source) {
      * @return the {@link InstanceType}
      */
     public static InstanceType fromHostSource(Map<String, Object> source) {
+        // Check and handle AWS.
+        String region = (String) source.get("ec2.placement.region");
+        if (region != null) {
+            String instanceType = (String) source.get("ec2.instance_type");
+            if (instanceType == null) {
+                instanceType = "";
+            }
+            return new InstanceType("aws", region, instanceType);
+        }
+
+        // Check and handle GCP.
+        String zone = (String) source.get("gce.instance.zone");
+        if (zone != null) {
+            // example: "gce.instance.zone": "projects/123456789/zones/europe-west1-b"
+            region = zone.substring(zone.lastIndexOf('/') + 1);
```

Comment:
All that parsing here makes me a bit nervous. To me that looks like a lot of complex code that interprets structured strings. Longer-term we should change our index structure so that we just need to read the respective properties (see my comment above).

Reply:
I agree. We make it part of the issue to be created (reference: https://github.com/elastic/elasticsearch/pull/103384#discussion_r1425339429).